SignalServiceKit/src/Messages/OWSMessageDecrypter.m

```diff
@@ -434,89 +435,137 @@ - (void)decryptUnidentifiedSender:(SSKProtoEnvelope *)envelope
+                    [self processException:underlyingException envelope:identifiedEnvelope];
+                    NSString *errorDescription = [NSString
+                        stringWithFormat:@"Exception while decrypting ud message: %@", underlyingException.description];
+                    OWSLogError(@"%@", errorDescription);
+                    NSError *error = OWSErrorWithCodeDescription(OWSErrorCodeFailedToDecryptMessage, errorDescription);
+                    failureBlock(error);
+                });
+                return;
             }
-            NSData *plaintextData = [decryptResult.paddedPayload removePadding];
-
-            SSKProtoEnvelopeBuilder *envelopeBuilder = [envelope asBuilder];
-            [envelopeBuilder setSource:source];
-            [envelopeBuilder setSourceDevice:(uint32_t)sourceDeviceId];
-            NSData *_Nullable newEnvelopeData = [envelopeBuilder buildSerializedDataAndReturnError:&error];
-            if (error || !newEnvelopeData) {
-                OWSFailDebug(@"Could not update UD envelope data: %@", error);
-                error = EnsureDecryptError(error, @"Could not update UD envelope data");
-                return failureBlock(error);
+
+            if ([underlyingError.domain isEqualToString:@"SignalMetadataKit.SMKSecretSessionCipherError"]
+                && underlyingError.code == SMKSecretSessionCipherErrorSelfSentMessage) {
+                // Self-sent messages can be safely discarded.
+                failureBlock(underlyingError);
+                return;
             }
 
-            OWSMessageDecryptResult *result = [OWSMessageDecryptResult resultWithEnvelopeData:newEnvelopeData
-                                                                                plaintextData:plaintextData
-                                                                                       source:source
-                                                                                 sourceDevice:(uint32_t)sourceDeviceId
-                                                                                  isUDMessage:YES];
-            successBlock(result, transaction);
-        } @catch (NSException *exception) {
-            dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
-                [self processException:exception envelope:envelope];
-                NSString *errorDescription =
-                    [NSString stringWithFormat:@"Exception while decrypting ud message: %@", exception.description];
-                OWSLogError(@"%@", errorDescription);
-                NSError *error = OWSErrorWithCodeDescription(OWSErrorCodeFailedToDecryptMessage, errorDescription);
-                failureBlock(error);
-            });
+            OWSFailDebug(@"Could not decrypt UD message: %@", underlyingError);
+            failureBlock(underlyingError);
+            return;
+        }
+
+        if (decryptResult.messageType == SMKMessageTypePrekey) {
```

Comment:
The remainder of the changes in the method are indent only, because it's no longer in a try/catch.